src/main/java/com/esri/samples/na/routing_around_barriers/RoutingAroundBarriersController.java

```diff
@@ -0,0 +1,335 @@
+
+          // add the new stop to the list of stops
+          stopsList.add(stopPoint);
+
+          // create a marker symbol and graphics, and add the graphics to the graphics overlay
+          CompositeSymbol newStopSymbol = createCompositeStopSymbol(stopsList.size());
+          Graphic stopGraphic = new Graphic(mapPoint, newStopSymbol);
+          stopsGraphicsOverlay.getGraphics().add(stopGraphic);
+
+        } else if (btnAddBarrier.isSelected()) {
+          // clear the displayed route, if it exists, since it might not be up to date any more
+          routeGraphicsOverlay.getGraphics().clear();
+
+          // create a buffered polygon around the mapPoint
+          Polygon bufferedBarrierPolygon = GeometryEngine.buffer(mapPoint, 500);
+
+          // create a polygon barrier for the routing task
+          PolygonBarrier barrier = new PolygonBarrier(bufferedBarrierPolygon);
+          barriersList.add(barrier);
+
+          // build graphics for the barrier and add it to the graphics overlay
+          Graphic barrierGraphic = new Graphic(bufferedBarrierPolygon, barrierSymbol);
+          barriersGraphicsOverlay.getGraphics().add(barrierGraphic);
+        }
+
+        // if the secondary mouse button was clicked, delete the last stop or barrier, respectively
+      } else if (e.getButton() == MouseButton.SECONDARY && e.isStillSincePress()) {
+        // clear the displayed route, if it exists, since it might not be up to date any more
+        routeGraphicsOverlay.getGraphics().clear();
+
+        if (btnAddStop.isSelected()) {
+          // check if there are stops to remove
+          if (!stopsList.isEmpty()) {
+
+            // remove the last stop from the stop list and the graphics overlay
+            stopsList.remove(stopsList.size() - 1);
+            stopsGraphicsOverlay.getGraphics().remove(stopsGraphicsOverlay.getGraphics().size() - 1);
+          }
+        } else if (btnAddBarrier.isSelected()) {
+          // check if there are barriers to remove
+          if (!barriersList.isEmpty()) {
+
+            // remove the last barrier from the barrier list and the graphics overlay
+            barriersList.remove(barriersList.size() - 1);
+            barriersGraphicsOverlay.getGraphics().remove(barriersGraphicsOverlay.getGraphics().size() - 1);
+          }
+        }
+      }
+    });
+  }
+
+  /**
+   * Uses the route task with the lists of stops and barriers to determine the route and display it.
+   */
+  @FXML
+  private void createRouteAndDisplay() {
+    if (stopsList.size() >= 2) {
+      // clear the previous route from the graphics overlay, if it exists
+      routeGraphicsOverlay.getGraphics().clear();
+
+      // add the existing stops and barriers to the route parameters
+      routeParameters.setStops(stopsList);
+      routeParameters.setPolygonBarriers(barriersList);
+
+      // apply the requested route finding parameters
+      routeParameters.setFindBestSequence(findBestSequenceCheckBox.isSelected());
```

Comment:
I don't think they do? If we ask the API to rearrange the stops (`find best sequence`), it will arrange all of them to make the shortest trip possible. Once we set that flag, the first or last stop aren't guaranteed to be in the same place any more. So after that we should get the option to fore the first or last stop to stay in that position.
I'm not sure if I understand what you are suggesting in the second part? 
Should remove the `findBestSequenceCheckBox`, and replace with just the `preserveFirstStopCheckBox` and `preserveLastStopCheckBox`, always enabled, and the `setFindBestSequence` only set to `true` once any of these boxes are ticked? If that's the case, then I don't think it'll be very clear to the user what is going on when they are ticking either of the boxes. Also, there's always the possibility the user want's the stops rearranged but doesn't insist on the first/last stops being preserved.

Reply:
Ok, I just looked at the API Ref. You're right. UI is correct.